### PR TITLE
solution was added

### PR DESCRIPTION
**Pull request by Maksym845**
- [DEMO LINK](https://Maksym845.github.io/react_todo-app-with-api/)

### CODE REVIEW

**Review by Veremchukk — APPROVED** [review]
GJ. But try to fix this problem

**Comment by Maksym845** [comment]
> GJ. But try to fix this problem

Yes, I forgot about this problem, thank you)

File: src/App.tsx [hunk 1]
```diff
@@ -1,18 +1,206 @@
...
+  const [updatingError, setUpdatingError] = useState(false);
+
+  const [toggleAll, setToggleAll] = useState(false);
+
+  const increaseCounter = () => {
+    setCounter(prevState => prevState + 1);
+  };
+
+  const filterTodos = (filterBy: FilterBy) => {
+    if (todos) {
```
Comment (Anna-Tkachenko) [inline src/App.tsx:50]: this check is redundant, you always have the link to array in todos variable

File: src/App.tsx [hunk 2]
```diff
@@ -1,18 +1,206 @@
...
+      setVisibleTodos(todos.filter((todo: Todo) => {
```
Comment (Anna-Tkachenko) [inline src/App.tsx:51]: use functional setState if you change it according to the current value

File: src/App.tsx [hunk 3]
```diff
@@ -1,18 +1,206 @@
...
+          default:
+            return true;
+        }
+      }));
+    }
+  };
+
+  const todosLength = () => {
+    return todos?.filter(todo => !todo.completed).length || '0';
+  };
```
Comment (Anna-Tkachenko) [inline src/App.tsx:66]: todos variable always store an array

File: src/App.tsx [hunk 4]
```diff
@@ -1,18 +1,206 @@
...
+
+  const todosCompletedLength = () => {
+    return todos?.filter(todo => todo.completed).length;
+  };
+
+  const getTodosFromApi = useCallback(async () => {
+    try {
+      if (user) {
+        setIsAdding(true);
+        const apiTodos = await getTodos(user.id).then(gotTodos => gotTodos);
```
Comment (Anna-Tkachenko) [inline src/App.tsx:76]: then after await function is redundant here

File: src/App.tsx [hunk 5]
```diff
@@ -1,18 +1,206 @@
...
+      setCounter((prevState) => prevState + 1);
+    }
+
+    setNewTodoTitle('');
+    setAddingNewTodo(false);
+  };
+
+  const onDeleteTodo = async (id: number) => {
+    await deleteTodo(id);
+    increaseCounter();
```
Comment (Anna-Tkachenko) [inline src/App.tsx:158]: for what purpose did you add counter value? can we calculate it according to the total length of todos?

File: src/App.tsx [hunk 6]
```diff
@@ -1,18 +1,206 @@
...
+  };
+
+  const onToggleAll = (newStatus: boolean) => {
+    setToggleAll(prevState => !prevState);
+    try {
+      todos.forEach(async (todo) => {
+        await updateTodo({
+          ...todo,
+          completed: newStatus,
+        }, todo.id);
```
Comment (Anna-Tkachenko) [inline src/App.tsx:183]: let's not change todos that are already completed / uncompleted

File: src/App.tsx [hunk 7]
```diff
@@ -23,201 +211,108 @@ export const App: React.FC = () => {
...
+              </button>
+            </footer>
+          </>
+        )}
       </div>
+
+      {(emptyTitleError
+        || addingError
+        || deletingError
+        || updatingError) && (
```
Comment (Anna-Tkachenko) [inline src/App.tsx:306]: let's create separate variable for storing the final result of this condition and use this variable in jsx

**Review by Anna-Tkachenko — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 8]
```diff
@@ -1,18 +1,201 @@
...
+  };
+
+  const onDeleteTodo = async (id: number) => {
+    await deleteTodo(id);
+    await getTodosFromApi();
+  };
+
+  const handleClearCompleted = () => {
+    setDeletingCompletedTodos(true);
+    todos?.forEach(async (todo) => {
```
Comment (Anna-Tkachenko) [inline src/App.tsx:151]: todos variable can't be undefined or null, so ? sign is redundant here

File: src/App.tsx [hunk 9]
```diff
@@ -23,201 +206,102 @@ export const App: React.FC = () => {
...
+            <TodosList
+              todos={visibleTodos}
+              onDeleteTodo={onDeleteTodo}
+              addingNewTodo={addingNewTodo}
+              tempTodo={{
+                id: 0,
+                userId: user.id,
+                title: newTodoTitle,
+                completed: false,
+              }}
```
Comment (Anna-Tkachenko) [inline src/App.tsx:240]: will be better to create separate variable for storing this object

**Review by Anna-Tkachenko — APPROVED** [review]
Good job! 🔥

**Pull request closed without merging**